# Installation  (Only Run Once)

Install sentence transformer.

Installation debug: https://chatgpt.com/share/67424793-a5b4-800f-8576-73f778bda7b2

In [ ]:
!pip install transformers sentence-transformers -q

Before we proceed, we need to ensure that the essential libraries are installed:
- `Hugging Face Transformers`: Provides us with a straightforward way to use pre-trained models.
- `PyTorch`: Serves as the backbone for deep learning operations.
- `Accelerate`: Optimizes PyTorch operations, especially on GPU.

In [ ]:
!pip install transformers torch accelerate

Defaulting to user installation because normal site-packages is not writeable


In [ ]:
pip install 'accelerate>=0.26.0'

Note: you may need to restart the kernel to use updated packages.


ERROR: Invalid requirement: "'accelerate": Expected package name at the start of dependency specifier
    'accelerate
    ^


# Import Libraries (Run Everytime)


Log in to Hugging Face

In [1]:
from huggingface_hub import login

login(token="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")

Load Llama-2 and tokenizer

In [2]:
from transformers import AutoTokenizer
import transformers
import torch

model = "meta-llama/Llama-2-7b-chat-hf" # meta-llama/Llama-2-7b-hf

tokenizer = AutoTokenizer.from_pretrained(model, use_auth_token=True)

/usr/local/lib/python3.10/dist-packages/transformers/models/auto/tokenization_auto.py:809: FutureWarning: The `use_auth_token` argument is deprecated and will be removed in v5 of Transformers. Please use `token` instead.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/1.62k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.84M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/414 [00:00<?, ?B/s]

In [3]:
from transformers import pipeline
from sentence_transformers import SentenceTransformer
import pandas as pd
import numpy as np

# Stage 1.1: Ask LLM Without RAG (Run When Needed):

## Set up a pipeline for text generation.

In [4]:
llama_pipeline = pipeline(
    "text-generation",  # LLM task
    model=model,
    torch_dtype=torch.float16,
    device_map=0,
)

config.json:   0%|          | 0.00/614 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/26.8k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.98G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/3.50G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/188 [00:00<?, ?B/s]

In [5]:
from accelerate import infer_auto_device_map, dispatch_model
from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline

# Load the model with low memory usage
model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-2-7b-chat-hf",
    device_map="auto",  # Automatically detect available devices
    offload_folder="offload",  # Folder to offload parts of the model to disk
    torch_dtype="auto",  # Automatically determine the appropriate precision
    offload_state_dict=True  # Enables state dict offloading to save memory
)

# Set up pipeline
tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-chat-hf")
llama_pipeline = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer,
    device_map="auto"
)

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [ ]:
PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True

SyntaxError: invalid syntax (<ipython-input-19-92cbf9cbbfb4>, line 1)

Set up function to prompt the LLM and get the generated text



In [6]:
def get_llama_response(prompt: str) -> None:
    sequences = llama_pipeline(
        prompt,
        do_sample=True,
        top_k=10,
        num_return_sequences=1,
        eos_token_id=tokenizer.eos_token_id,
        max_length=256,
        truncation=True,
    )
    print("Chatbot:", sequences[0]['generated_text'])

## Tests and Results:

The description of the ingredient in the document


> Ascorbyl Glucoside: Part of the Vitamin C family of ingredients, Ascorbyl Glucoside is a water-soluble derivative of Vitamin C. This derivative was created to provide improved stability and skin penetration vs. ascorbic acid alone. It is still unclear how well ascorbyl glucoside is converted into ascorbic acid (the active form of Vitamin C) within the body, but some of it will be lost during this conversion, making this less potent, and therefore requiring higher levels in the formulation in order to achieve great results.
Vitamin C can also be found on ingredients lists as ascorbic acid, ascorbyl palmitate, ethyl ascorbic acid, l-ascorbic acid, magnesium ascorbyl phosphate, sodium ascorbyl phosphate and tetrahexyldecyl ascorbate.. Ascorbyl Glucoside is a water-soluble derivative of Vitamin C.. It is used in skin care to brighten the skin and improve the signs of aging.
 Ascorbyl Glucoside is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Wrinkles.
 People with Related Allergy, Impaired skin barrier should avoid Ascorbyl Glucoside.

Below are answer from Llama-2 without the help of the knowledge document.

In [ ]:
query = "What is Ascorbyl Glucoside good for?"
get_llama_response(query)

Chatbot: What is Ascorbyl Glucoside good for?
Ascorbyl glucoside is a stable and bioavailable form of vitamin C that has been shown to have a range of benefits for skin health. Here are some of the potential benefits of ascorbyl glucoside:
1. Antioxidant activity: Ascorbyl glucoside has been shown to have antioxidant activity, which can help protect the skin from damage caused by free radicals.
2. Collagen production: Vitamin C is important for the production of collagen, a protein that gives skin its strength and elasticity. Ascorbyl glucoside may help to boost collagen production in the skin.
3. Hydration: Ascorbyl glucoside can help to improve skin hydration by increasing the amount of water in the skin.
4. Brightening and even skin tone: Ascorbyl glucoside may help to brighten and even out skin tone by inhibiting the formation of melanin.
5. Anti-inflammatory activity: Ascorbyl glucoside has been shown to have


In [ ]:
query = "Is Ascorbyl Glucoside good for wrinkles?"
get_llama_response(query)

Chatbot: Is Ascorbyl Glucoside good for wrinkles?

Ascorbyl glucoside is a stable and bioavailable form of vitamin C that has been shown to have a number of benefits for the skin, including reducing the appearance of wrinkles. Here are some reasons why ascorbyl glucoside may be good for wrinkles:

1. Hydration: Ascorbyl glucoside helps to improve skin hydration by increasing the amount of hyaluronic acid in the skin. Hyaluronic acid is a natural moisturizer that can help to plump up the skin and reduce the appearance of fine lines and wrinkles.
2. Collagen production: Ascorbyl glucoside can help to stimulate collagen production in the skin, which can help to improve skin elasticity and reduce the appearance of wrinkles.
3. Antioxidant protection: Ascorbyl glucoside has antioxidant properties, which can help to protect the skin from damage caused by free radicals. This can help to reduce the appearance of wrinkles and improve the overall health and


In [ ]:
query = "Can someone with impaired skin barrier use Ascorbyl Glucoside?"
get_llama_response(query)

Chatbot: Can someone with impaired skin barrier use Ascorbyl Glucoside?

Ascorbyl Glucoside is a stable and bioavailable form of Vitamin C that can help to improve the skin's barrier function. However, if someone has impaired skin barrier, they may need to take extra precautions when using Ascorbyl Glucoside to avoid further irritating or compromising the skin's natural barrier function. Here are some tips for using Ascorbyl Glucoside with impaired skin barrier:

1. Start with a low concentration: If you have impaired skin barrier, it's best to start with a lower concentration of Ascorbyl Glucoside and gradually increase as your skin becomes more tolerant.
2. Use it in the evening: Ascorbyl Glucoside can be more irritating to the skin when used in the morning, as the skin's natural barrier function is weaker during this time. Using it in the evening can help to minimize any potential irritation.
3. Use a moisturizer: Applying a moistur


Use the function to create an interactive chat loop, where the user can converse with the Llama model.

In [ ]:
while True:
    user_input = input("You: ")
    if user_input.lower() in ["bye", "quit", "exit"]:
        print("Chatbot: Goodbye!")
        break
    get_llama_response(user_input)

You: Is Ascorbic Acid good for dry skin?


Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.


Chatbot: Is Ascorbic Acid good for dry skin?

Ascorbic Acid, also known as Vitamin C, is a popular ingredient in skincare products that can provide several benefits for dry skin. Here are some reasons why Ascorbic Acid can be good for dry skin:

1. Hydration: Ascorbic Acid can help to hydrate dry skin by increasing the production of collagen, which is an essential protein for skin health. Collagen helps to plump up the skin, giving it a more youthful and radiant appearance.
2. Anti-inflammatory properties: Ascorbic Acid has anti-inflammatory properties that can help to reduce redness and irritation in dry skin. This can be especially beneficial for people who suffer from conditions like eczema or psoriasis.
3. Exfoliating properties: Ascorbic Acid can help to exfoliate dry skin, removing dead skin cells and revealing smoother, brighter skin. This can help to improve skin texture and reduce the appearance of fine lines and wrinkles.
4. Antioxidant properties
You: Is Ethyl Ascorbic Acid 

We should ask a wider range of questions.

The set of questions we asked and answer based on the knowledge document:
*   Is Ascorbic Acid good for dry skin? (No)
*   Is Ethyl Ascorbic Acid bad for impaired skin barrier? (Yes)

*   What are the benefits of Glycoproteins? (Essential lipids offer benefits such as: - Help strengthen the skin's natural barrier, improving it...)
*   Who should avoid Mandelic Acid?('Related Allergy', ' ', 'Sensitive', ' ', 'Impaired skin barrier')
*   I want to reduce wrinkles. What should I use?







##Evaluation

In general, the Llama-2 model answer decently without referencing to the knowledge base. However, there are issues such as

*   not answering the question in the context of skincare (e.g., answer to "What are the benefits of Glycoproteins?": it answered in a genreal way)
*   not giving the right answer (e.g., answer to "Is Ascorbic Acid good for dry skin?": while it said yes, the expert actually suggest people with dry skin to avoid it)

Therefore, we think incorporating RAG is still valuable.



# Stage 1.2 Basic RAG

## Data Preprocessing (Don't Run)

In [41]:
import pandas as pd
import numpy as np

In [74]:
df=pd.read_csv('ingredientsList.csv')
df

name scientific_name  \
0    Alpha-Glucan Oligosaccharide             NaN   
1                       Aloe Vera             NaN   
2                       Allantoin             NaN   
3                           Algin             NaN   
4                   Algae Extract             NaN   
..                            ...             ...   
243                  Zinc Sulfate             NaN   
244                      Zinc PCA             NaN   
245                Neoglucosamine             NaN   
246                       Taurine             NaN   
247     Sodium Ascorbyl Phosphate             NaN   

                                     short_description  \
0    Alpha-glucan oligosaccharide is in a class of ...   
1    Aloe vera, also appear on ingredients lists as...   
2    Allantoin occurs naturally in the body, but ca...   
3    Algin, also known as sodium alginate, is a lar...   
4    It is essentially an underwater plant, designe...   
..                                                 ...   
243  The jury is still out on the efficacy of this ...   
244  Zinc PCA is a component of the skin’s natural ...   
245  A patented non-amino acid sugar which gently e...   
246  In skin care, taurine is a powerful form of an...   
247  Sodium Ascorbyl Phosphate (SAP) is a gentle an...   

                                            what_is_it  \
0    Prebiotics are a type of non-digestible dietar...   
1    Aloe vera is a skincare ingredient derived fro...   
2    Allantoin is a skincare ingredient derived fro...   
3    An extract from brown seaweed used as hydratin...   
4    An incredibly interesting natural ingredient s...   
..                                                 ...   
243      An antibacterial ingredient used in skincare.   
244  Zinc PCA is known to regulate sebum production...   
245  Neoglucosamine is a patented skincare ingredie...   
246  Taurine is an amino acid that’s naturally prod...   
247        A gentle, water soluble form of Vitamin C.    

                                       what_does_it_do  \
0    Prebiotics offer benefits such as:\r\n\r\n- He...   
1    Aloe vera offers multiple benefits for the ski...   
2    Allantoin is a calming, anti-inflammatory, and...   
3    In skincare products, it is used for its excel...   
4    Algae extracts are multifunctional ingredients...   
..                                                 ...   
243  Zinc salts such as zinc gluconate, zinc PCA an...   
244  Zinc salts such as zinc gluconate, zinc PCA an...   
245  Neoglucosamine offers benefits such as:\n- Sup...   
246  Helps support the skin barrier and protect aga...   
247  Sodium ascorbyl phosphate helps brighten the s...   

                                    who_is_it_good_for  \
0    [' ', 'Acne', ' ', 'Blackheads', ' ', 'Redness...   
1    [' ', 'Dry and dehydrated skin', ' ', 'Impaire...   
2    [' ', 'Fine Lines', ' ', 'Impaired skin barrie...   
3    [' ', 'Dry and dehydrated skin', ' ', 'Fine Li...   
4    [' ', 'Dry and dehydrated skin', ' ', 'Fine Li...   
..                                                 ...   
243                                    [' ', 'Anyone']   
244  [' ', 'Acne', ' ', 'Blackheads', ' ', 'Enlarge...   
245  [' ', 'Blackheads', ' ', 'Elasticity', ' ', 'E...   
246  [' ', 'Dry and dehydrated skin', ' ', 'Radianc...   
247  [' ', 'Acne', ' ', 'Blackheads', ' ', 'Fine Li...   

                                      who_should_avoid  \
0                             [' ', 'Related Allergy']   
1                             [' ', 'Related Allergy']   
2                             [' ', 'Related Allergy']   
3                             [' ', 'Related Allergy']   
4                             [' ', 'Related Allergy']   
..                                                 ...   
243                           [' ', 'Related Allergy']   
244                           [' ', 'Related Allergy']   
245  [' ', 'Related Allergy', ' ', 'Impaired skin b...   
246                     

Clean up the data sheet by formating things into sentences and removing extra punctuation.

In [75]:
import ast
# Function to clean up and convert the "who_is_it_good_for" column into a sentence
def process_good_for(name,value):
    # Convert the string representation of a list into an actual list
    items = ast.literal_eval(value)
    # Remove empty strings or strings with only spaces
    items = [item.strip() for item in items if item.strip()]
    # Join the cleaned items into a sentence
    if items:
        return f"{name} is suitable for: " + ", ".join(items) + "."
    return ""

# Function to clean and join the "who_should_avoid" column
def process_bad_for(name, value):
     # Convert the string representation of a list into an actual list
    items = ast.literal_eval(value)
    # Remove empty strings or strings with only spaces
    items = [item.strip() for item in items if item.strip()]
    # Join the cleaned items into a sentence
    if items:
        return f"People with {','.join(items)} should avoid {name}."
    return ""

# Join all the columns into one and make new knowledge base
df['combined'] = df.apply(lambda row: (
    f" {process_good_for(row['name'], row['who_is_it_good_for'])}\n"
    f" {process_bad_for(row['name'], row['who_should_avoid'])}\n"
    f"{row['name']}: {row['what_does_it_do']}"
), axis=1)
print(df['combined'][0])

 Alpha-Glucan Oligosaccharide is suitable for: Acne, Blackheads, Redness, Pregnancy.
 People with Related Allergy should avoid Alpha-Glucan Oligosaccharide.
Alpha-Glucan Oligosaccharide: Prebiotics offer benefits such as:

- Help maintain a balanced and healthy skin microbiome by promoting the growth of beneficial bacteria.

- Support the skin's natural defences against harmful microorganisms, contributing to overall skin health.

- Enhance the skin's barrier function, reducing moisture loss and improving hydration.

- Can have a calming effect on the skin, helping to soothe and alleviate inflammation and irritation.


In [76]:
df.drop (['scientific_name','what_is_it', 'short_description', 'url', 'what_does_it_do'], axis=1, inplace=True)

In [77]:
#Create a new file
df.to_csv('ingredientList_processed.csv', index=False)

## Generate Dense Embeddings and Build an Index

In [ ]:
from sentence_transformers import SentenceTransformer

In [47]:
#Step 1: Read the documents (preprocessed text)
documents=df['combined']

# Step 2: Create dense embeddings for the documents
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')  # A lightweight model for embedding
embeddings = embedding_model.encode(documents)  # Generate embeddings

# Step 3: Create a pandas DataFrame for indexing
# Combine documents and embeddings in a DataFrame
df = pd.DataFrame({
    "Document": documents,
    "Embedding": list(embeddings)  # Store embeddings as lists for easier handling
})

# Display the DataFrame
print("Indexed Documents:")
print(df.head())

Indexed Documents:
                                            Document  \
0  Alpha-Glucan Oligosaccharide: Prebiotics offer...   
1  Aloe Vera: Aloe vera offers multiple benefits ...   
2  Allantoin: Allantoin is a calming, anti-inflam...   
3  Algin: In skincare products, it is used for it...   
4  Algae Extract: Algae extracts are multifunctio...   

                                           Embedding  
0  [-0.0011034776, 0.0107276365, 0.0009673794, 0....  
1  [0.02319691, 0.0886462, 0.072152, 0.07917443, ...  
2  [-0.056520388, -0.02964613, 0.0045101917, 0.06...  
3  [-0.08027446, -0.03375384, 0.024516206, 0.0196...  
4  [-0.09272464, 0.061912265, -0.009909179, 0.048...  


In [48]:
df

Document  \
0    Alpha-Glucan Oligosaccharide: Prebiotics offer...   
1    Aloe Vera: Aloe vera offers multiple benefits ...   
2    Allantoin: Allantoin is a calming, anti-inflam...   
3    Algin: In skincare products, it is used for it...   
4    Algae Extract: Algae extracts are multifunctio...   
..                                                 ...   
243  Zinc Sulfate: Zinc salts such as zinc gluconat...   
244  Zinc PCA: Zinc salts such as zinc gluconate, z...   
245  Neoglucosamine: Neoglucosamine offers benefits...   
246  Taurine: Helps support the skin barrier and pr...   
247  Sodium Ascorbyl Phosphate: Sodium ascorbyl pho...   

                                             Embedding  
0    [-0.0011034776, 0.0107276365, 0.0009673794, 0....  
1    [0.02319691, 0.0886462, 0.072152, 0.07917443, ...  
2    [-0.056520388, -0.02964613, 0.0045101917, 0.06...  
3    [-0.08027446, -0.03375384, 0.024516206, 0.0196...  
4    [-0.09272464, 0.061912265, -0.009909179, 0.048...  
..                                                 ...  
243  [-0.059526738, 0.020399818, -0.021279808, -0.0...  
244  [-0.07595724, 0.041505683, -0.04451312, -0.000...  
245  [-0.03784679, 0.04927488, -0.027302664, 0.0278...  
246  [-0.006863437, -0.040219788, 0.0859309, -0.013...  
247  [-0.0951243, 0.0022773868, 0.026917184, 0.0224...  

[248 rows x 2 columns]

In [49]:
#Create a new file to save the embedding
df.to_csv('knowledge_embedding.csv')

## Build the Retrival Component

In [ ]:
import pandas as pd
import numpy as np


from sentence_transformers import SentenceTransformer
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')  # A lightweight model for embedding

/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:13: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [61]:
# Load the knowledage data
knowledge_df=pd.read_csv('knowledge_embedding_fixed.csv')

# Convert string embeddings to lists
import ast
knowledge_df['Embedding'] = knowledge_df['Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split()],
        dtype=np.float32
    )
)

print(type( knowledge_df['Embedding'][0]))

<class 'numpy.ndarray'>


In [62]:
#function to retrieve knowledge
def retrieve_with_pandas(query, top_k=2):
    # Generate embedding for the query
    query_embedding = embedding_model.encode([query])[0]

    # Compute similarity scores (cosine similarity)
    knowledge_df['Similarity'] =  knowledge_df['Embedding'].apply(lambda x: np.dot(query_embedding, x) /
                                             (np.linalg.norm(query_embedding) * np.linalg.norm(x)))
    # Sort by similarity and return top-k results
    results =  knowledge_df.sort_values(by="Similarity", ascending=False).head(top_k)
    print("\nTop Retrieved Documents:") # for debugging

    return results[["Document", "Similarity"]]

### Retrival Evaluation 1

test if the retrieve funtion works fine

In [63]:
# Retrieve based on a question
query  = "Is Ascorbic Acid good for dry skin?"
results = retrieve_with_pandas(query, top_k=2)

print("\nQuery:", query)
print("\nTop Retrieved Documents:")
print(results)


Top Retrieved Documents:

Query: Is Ascorbic Acid good for dry skin?

Top Retrieved Documents:
                                              Document  Similarity
246  Sodium Ascorbyl Phosphate: Sodium Ascorbyl Pho...    0.739062
15   Ascorbic Acid: Part of the Vitamin C family of...    0.705302


**Qualitative Analysis:**
*   It's not retriving the right answer.
The retrival method needs improvement.
*   Assumption of the cause of the problem:  We made the embedding from the whole document, which is very long and can have extra noises.
*   Next step for this: use techniques (Dual - Encoder)  -> generate embedding for
smth else to calculate the score and then find the coresopnding document for text genreation.
*   (!!! Need to write this in the right terms for the report)

Now let's do a quantitative test.

Generate and record embeddings for the test queries.

In [ ]:
# Run only once!!!!

embedding_model = SentenceTransformer('all-MiniLM-L6-v2')

# Generate embeddings for the test queries
test_df["Query_Embedding"] = test_df["User_Question"].apply(
    lambda query: embedding_model.encode([query])[0].tolist()
)

# Save updated test_df to file
test_df.to_csv("Retrival_Test_Set_Embedding.csv", index=False)
print("Test query embeddings saved to 'Retrival_Test_Embedding.csv'.")

Test query embeddings saved to 'Retrival_Test_Embedding.csv'.


Use Saved Embeddings for Retrieval

In [65]:
import ast  # Import ast to handle string representations of lists

test_df = pd.read_csv("Retrival_Test_Set_Embedding.csv")

# Convert string embeddings to lists
test_df['Query_Embedding'] = test_df['Query_Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split(",")],
        dtype=np.float32
    )
)

knowledge_df = pd.read_csv("knowledge_embedding_fixed.csv")
knowledge_df['Embedding'] = knowledge_df['Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split()],
        dtype=np.float32
    )
)

In [67]:
retrieval_results = []

for _, row in test_df.iterrows():
    query_embedding = np.array(row["Query_Embedding"])  # Load query embedding
    ground_truth_index = row["Index_Ingredient_Name"] # The ground truth index

    # Compute similarity and retrieve top-2 documents
    knowledge_df['Similarity'] = knowledge_df['Embedding'].apply(
        lambda x: np.dot(query_embedding, x) / # Directly use x, no need for ast.literal_eval
                  (np.linalg.norm(query_embedding) * np.linalg.norm(x))  # Directly use x, no need for ast.literal_eval
    )
    top_k_indices = knowledge_df.sort_values(by="Similarity", ascending=False).head(2)["Index"].tolist()

    # Check if ground truth is in the top-2 and top-1
    is_in_top_2 = int(ground_truth_index in top_k_indices)
    is_in_top_1 = int(ground_truth_index == top_k_indices[0])

    # Save results
    retrieval_results.append({
        "Query": row["User_Question"],
        "Document_Retrived_1": knowledge_df.loc[top_k_indices[0], "Document"],
        "Document_Retrived_2": knowledge_df.loc[top_k_indices[1], "Document"],
        "Index_Document_Ground_Truth": ground_truth_index,
        "Indexes_Document_Retrived": top_k_indices,
        "Is_Ground_Truth_In_Top_2": is_in_top_2,
        "Is_Ground_Truth_In_Top_1": is_in_top_1
    })

# Convert results to DataFrame and save as CSV
results_df = pd.DataFrame(retrieval_results)
results_df.to_csv("retrival_result_document.csv", index=False)
print("Retrieval results saved to 'retrival_result_document_fixed.csv'.")

Retrieval results saved to 'retrival_result_document_fixed.csv'.


**Calculate Precision, Recall, and F1 Score**

In [68]:
from sklearn.metrics import precision_score, recall_score, f1_score

# Extract true labels and predictions for both Top-2 and Top-1
y_true_top_2 = results_df["Is_Ground_Truth_In_Top_2"]
y_pred_top_2 = [1] * len(results_df)  # Since we always retrieve 2 documents, predictions are always 1

y_true_top_1 = results_df["Is_Ground_Truth_In_Top_1"]
y_pred_top_1 = [1] * len(results_df)  # Predictions are 1 as we evaluate Top-1 retrieval.

# Metrics for Top-2
precision_top_2 = precision_score(y_true_top_2, y_pred_top_2, zero_division=1)
recall_top_2 = recall_score(y_true_top_2, y_pred_top_2, zero_division=1)
f1_top_2 = f1_score(y_true_top_2, y_pred_top_2, zero_division=1)

# Metrics for Top-1
precision_top_1 = precision_score(y_true_top_1, y_pred_top_1, zero_division=1)
recall_top_1 = recall_score(y_true_top_1, y_pred_top_1, zero_division=1)
f1_top_1 = f1_score(y_true_top_1, y_pred_top_1, zero_division=1)

# Print results
print(f"Top-2 Precision: {precision_top_2}, Recall: {recall_top_2}, F1 Score: {f1_top_2}")
print(f"Top-1 Precision: {precision_top_1}, Recall: {recall_top_1}, F1 Score: {f1_top_1}")

Top-2 Precision: 0.8117408906882592, Recall: 1.0, F1 Score: 0.8960893854748604
Top-1 Precision: 0.7591093117408907, Recall: 1.0, F1 Score: 0.8630609896432682


**Analysis:** Decent numbers, but have room for improvement.

##Build the Generation component

### 1.2.1 Try with the tutorial model first (Don't Need to Run)

In [ ]:
llm = pipeline('text-generation', model='gpt2')  # A lightweight example for demonstration

Hardware accelerator e.g. GPU is available in the environment, but no `device` argument is passed to the `Pipeline` object. Model will be on CPU.


In [ ]:
gpt2_pipeline = pipeline(
    "text-generation",  # LLM task
    model='gpt2',       # A lightweight model used in code from lab
    torch_dtype=torch.float16,
    device_map=0,
)

In [ ]:
# function to use RAG (GPT2 is the pre-trained LLM)
def generate_with_rag_GPT2(query):
    # Retrieve context
    context = retrieve_with_pandas(query)["Document"]
    # Combine context with query
    input_text = f"Respond the query based on the context. Context: {context} Query: {query}"
    # Use LLM to generate a response
    response = gpt2_pipeline(
        input_text,
        do_sample=True,
        top_k=10,
        num_return_sequences=1,
        max_length=256,
        truncation=True,
        )
    return response[0]['generated_text']

In [ ]:
query  = "Is Ascorbic Acid good for dry skin?"
results = generate_with_rag(query)

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.



Top Retrieved Documents:


In [ ]:
# function to use RAG (GPT2 is the pre-trained LLM)
def generate_with_rag(query):
    # Retrieve context
    context = retrieve_with_pandas(query)["Document"]
    # Combine context with query
    input_text = f"Respond the query based on the context. Context: {context} Query: {query}"
    # Use LLM to generate a response
    response = llm(input_text, max_length=150, num_return_sequences=1)

    #save answet in a csv for debug checking
    # Extract the 'generated_text' values into a list
    texts = [item['generated_text'] for item in response]
    # Create the DataFrame and save it
    response_df = pd.DataFrame({'Generated Text': texts})
    response_df.to_csv('temp.csv')

    return response[0]['generated_text']


#### Test Run

In [ ]:
query  = "Is Ascorbic Acid good for dry skin?"
response= generate_with_rag(query)
response

Output:

> 'Respond the query based on the context. Context: 247    Sodium Ascorbyl Phosphate: Sodium Ascorbyl Pho...\n15     Ascorbic Acid: Part of the Vitamin C family of...\nName: Document, dtype: object Query: Is Ascorbic Acid good for dry skin? Result: Yes Name: Ascorbic Acid - Sodium Ascorbyl Phosphate- ascorbic acid, - Sodium Ascorbico...\nName: Document, dtype: object Query: Is Ascorbic Acid bad for dry skin? Result: Yes Name: Ascorbic Acid - Sodium Ascorbyl Phosphate- as'



####Result and Evaluation

**Result:**
> Respond the query based on the context. Context: 247    Sodium Ascorbyl Phosphate: Sodium Ascorbyl Pho...
15     Ascorbic Acid: Part of the Vitamin C family of...
Name: Document, dtype: object Query: Is Ascorbic Acid good for dry skin?

> Result: Yes Name: Ascorbic Acid - Sodium Ascorbyl Phosphate- ascorbic acid, - Sodium Ascorbico...
Name: Document, dtype: object Query: Is Ascorbic Acid bad for dry skin? Result: Yes Name: Ascorbic Acid - Sodium Ascorbyl Phosphate- as

**Evaluation:**

The output doesn't really make sense. Maybe the GPT2 model is not working well. We should try a more powerful LLM.

### 1.2.2 Replace GPT2 with Llama2

In [ ]:
llama_pipeline = pipeline(
    "text-generation",  # LLM task
    model=model,
    torch_dtype=torch.float16,
    device_map=0,
)

config.json:   0%|          | 0.00/614 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/26.8k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.98G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/3.50G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/188 [00:00<?, ?B/s]

In [ ]:
#function to use RAG (Llama is the pre-trained LLM)
def generate_with_rag_Llama(query):
    # Retrieve context
    context = retrieve_with_pandas(query)["Document"]
    # Combine context with query
    input_text = f"Respond the query based on the context. Context: {context} Query: {query}"
    # Use LLM to generate a response
    response = llama_pipeline(
        input_text,
        do_sample=True,
        top_k=10,
        num_return_sequences=1,
        eos_token_id=tokenizer.eos_token_id,
        max_length=256,
        truncation=True
        )
    return response[0]['generated_text']


####Test, Result, and Evaluation

In [ ]:
query  = "Is Ascorbic Acid good for dry skin?"
response= generate_with_rag_Llama(query)
response


Top Retrieved Documents:


"Respond the query based on the context. Context: 247    Sodium Ascorbyl Phosphate: Sodium Ascorbyl Pho...\n15     Ascorbic Acid: Part of the Vitamin C family of...\nName: Document, dtype: object Query: Is Ascorbic Acid good for dry skin?\n\nAscorbic Acid, also known as Vitamin C, is a powerful antioxidant that can help to combat dry skin. It is an essential nutrient that plays a crucial role in the production of collagen, which is an important protein that helps to keep skin firm and elastic.\n\nWhen the skin loses its natural moisture, it can become dry and flaky, leading to discomfort and irritation. Ascorbic Acid can help to combat this by increasing the skin's natural moisture barrier, reducing inflammation and redness, and promoting the growth of new skin cells.\n\nIn addition, Ascorbic Acid has antibacterial properties, which can help to prevent bacterial infections that can exacerbate dry skin"

Llama-2 did slightly better than GPT-2. But the output is still weird.

# Stage 1 Coclusion and Next Steps
Attempt 1.2 (using rag) is worse than attempt 1.1 (using LLM without RAG). The main issues are (1)bad retrival, (2) unoptimized prompt engineering. In the following attempts, we will focuse on improving the retrival and generation components:
- Revise the RAG retrival (Attempt 2):
  - Option 1 : use questions to generate embedding.
  - Option 2 : element name to generate embedding.
- Improve Generation (Attempt 3):
  - Option 1: Prompt engineering
  - Option 2: Fine tune Llama 2.

**We didn't do a comprehensive evaluation in this first stage because we just wanted a basic understanding of how the models and components are working and. issues were easy to identify at this stage.**



# Stage 2.1: Change RAG Retrival Strategy - retrive with ingredient name

## Data Preprocessing - Embedding of Ingredient Names (Don't Run)


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df=pd.read_csv('ingredientList_processed.csv')
df

name  \
0    Alpha-Glucan Oligosaccharide   
1                       Aloe Vera   
2                       Allantoin   
3                           Algin   
4                   Algae Extract   
..                            ...   
243                  Zinc Sulfate   
244                      Zinc PCA   
245                Neoglucosamine   
246                       Taurine   
247     Sodium Ascorbyl Phosphate   

                                    who_is_it_good_for  \
0    [' ', 'Acne', ' ', 'Blackheads', ' ', 'Redness...   
1    [' ', 'Dry and dehydrated skin', ' ', 'Impaire...   
2    [' ', 'Fine Lines', ' ', 'Impaired skin barrie...   
3    [' ', 'Dry and dehydrated skin', ' ', 'Fine Li...   
4    [' ', 'Dry and dehydrated skin', ' ', 'Fine Li...   
..                                                 ...   
243                                    [' ', 'Anyone']   
244  [' ', 'Acne', ' ', 'Blackheads', ' ', 'Enlarge...   
245  [' ', 'Blackheads', ' ', 'Elasticity', ' ', 'E...   
246  [' ', 'Dry and dehydrated skin', ' ', 'Radianc...   
247  [' ', 'Acne', ' ', 'Blackheads', ' ', 'Fine Li...   

                                      who_should_avoid  \
0                             [' ', 'Related Allergy']   
1                             [' ', 'Related Allergy']   
2                             [' ', 'Related Allergy']   
3                             [' ', 'Related Allergy']   
4                             [' ', 'Related Allergy']   
..                                                 ...   
243                           [' ', 'Related Allergy']   
244                           [' ', 'Related Allergy']   
245  [' ', 'Related Allergy', ' ', 'Impaired skin b...   
246                           [' ', 'Related Allergy']   
247  [' ', 'Related Allergy', ' ', 'Impaired skin b...   

                                              combined  
0    Alpha-Glucan Oligosaccharide: Alpha-glucan oli...  
1    Aloe Vera: Aloe vera, also appear on ingredien...  
2    Allantoin: Allantoin occurs naturally in the b...  
3    Algin: Algin, also known as sodium alginate, i...  
4    Algae Extract: It is essentially an underwater...  
..                                                 ...  
243  Zinc Sulfate: The jury is still out on the eff...  
244  Zinc PCA: Zinc PCA is a component of the skin’...  
245  Neoglucosamine: A patented non-amino acid suga...  
246  Taurine: In skin care, taurine is a powerful f...  
247  Sodium Ascorbyl Phosphate: Sodium Ascorbyl Pho...  

[248 rows x 4 columns]

## Generate Dense Embeddings and Build an Index

In [ ]:
from sentence_transformers import SentenceTransformer

In [ ]:
#Step 1: Read the documents (preprocessed text)
documents=df['combined']
# Read ingredient names
ingredient_names = df['name']

# Step 2: Create dense embeddings for the 'names' instead of 'documents'
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')  # A lightweight model for embedding
embeddings = embedding_model.encode(ingredient_names)  # Generate embeddings

# Step 3: Create a pandas DataFrame for indexing
# Combine documents and embeddings in a DataFrame
df = pd.DataFrame({
    "Ingredient": ingredient_names,
    "Document": documents,
    "Embedding": list(embeddings)  # Store embeddings as lists for easier handling
})

# Display the DataFrame
print("Indexed Documents:")
print(df.head())

Indexed Documents:
                     Ingredient  \
0  Alpha-Glucan Oligosaccharide   
1                     Aloe Vera   
2                     Allantoin   
3                         Algin   
4                 Algae Extract   

                                            Document  \
0  Alpha-Glucan Oligosaccharide: Alpha-glucan oli...   
1  Aloe Vera: Aloe vera, also appear on ingredien...   
2  Allantoin: Allantoin occurs naturally in the b...   
3  Algin: Algin, also known as sodium alginate, i...   
4  Algae Extract: It is essentially an underwater...   

                                           Embedding  
0  [-0.0018100978, -0.055112716, -0.08810373, 0.0...  
1  [-0.0043009343, 0.05342271, 0.027977332, 0.062...  
2  [-0.090445615, 0.011692092, -0.039966017, 0.02...  
3  [-0.0710434, -0.013296724, 0.012745387, 0.0286...  
4  [-0.10274659, 0.0737859, -0.0634238, 0.0335935...  


In [ ]:
df

Ingredient  \
0    Alpha-Glucan Oligosaccharide   
1                       Aloe Vera   
2                       Allantoin   
3                           Algin   
4                   Algae Extract   
..                            ...   
243                  Zinc Sulfate   
244                      Zinc PCA   
245                Neoglucosamine   
246                       Taurine   
247     Sodium Ascorbyl Phosphate   

                                              Document  \
0    Alpha-Glucan Oligosaccharide: Alpha-glucan oli...   
1    Aloe Vera: Aloe vera, also appear on ingredien...   
2    Allantoin: Allantoin occurs naturally in the b...   
3    Algin: Algin, also known as sodium alginate, i...   
4    Algae Extract: It is essentially an underwater...   
..                                                 ...   
243  Zinc Sulfate: The jury is still out on the eff...   
244  Zinc PCA: Zinc PCA is a component of the skin’...   
245  Neoglucosamine: A patented non-amino acid suga...   
246  Taurine: In skin care, taurine is a powerful f...   
247  Sodium Ascorbyl Phosphate: Sodium Ascorbyl Pho...   

                                             Embedding  
0    [-0.0018100978, -0.055112716, -0.08810373, 0.0...  
1    [-0.0043009343, 0.05342271, 0.027977332, 0.062...  
2    [-0.090445615, 0.011692092, -0.039966017, 0.02...  
3    [-0.0710434, -0.013296724, 0.012745387, 0.0286...  
4    [-0.10274659, 0.0737859, -0.0634238, 0.0335935...  
..                                                 ...  
243  [-0.14192155, 0.05794519, -0.02086188, 0.00191...  
244  [-0.15434268, 0.07663269, -0.06987211, 0.01624...  
245  [-0.013301337, 0.008945664, -0.089915715, 0.01...  
246  [-0.004846669, -0.0067391316, 0.013697509, -0....  
247  [-0.1126298, -0.041794844, -0.05042698, 0.0342...  

[248 rows x 3 columns]

In [ ]:
#Create a new file to save the embedding
df.to_csv('knowledge_embedding_name.csv')

## Retrival Evaluation 2.1

In [ ]:
# Load the knowledage data
knowledge_df=pd.read_csv('knowledge_embedding_name.csv')

# Convert string embeddings to lists
import ast
knowledge_df['Embedding'] = knowledge_df['Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split()],
        dtype=np.float32
    )
)

print(type( knowledge_df['Embedding'][0]))

<class 'numpy.ndarray'>


In [ ]:
# Retrieve based on a question
query  = "Is Ascorbic Acid good for dry skin?"
results = retrieve_with_pandas(query, top_k=2)

print("\nQuery:", query)
print("\nTop Retrieved Documents:")
print(results)


Top Retrieved Documents:

Query: Is Ascorbic Acid good for dry skin?

Top Retrieved Documents:
                                             Document  Similarity
15  Ascorbic Acid: Part of the Vitamin C family of...    0.692339
89  Ethyl Ascorbic Acid: Part of the Vitamin C fam...    0.631519


**Qualitative Analysis:** This time it retrived the right one!

Now let's do a quantitative test.

Use Saved Embeddings for Retrieval

In [ ]:
import ast  # Import ast to handle string representations of lists

test_df = pd.read_csv("Retrival_Test_Set_Embedding.csv")

# Convert string embeddings to lists
test_df['Query_Embedding'] = test_df['Query_Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split(",")],
        dtype=np.float32
    )
)

In [ ]:
retrieval_results = []

for _, row in test_df.iterrows():
    query_embedding = np.array(row["Query_Embedding"])  # Load query embedding
    ground_truth_index = row["Index_Ingredient_Name"] # The ground truth index

    # Compute similarity and retrieve top-2 documents
    knowledge_df['Similarity'] = knowledge_df['Embedding'].apply(
        lambda x: np.dot(query_embedding, x) / # Directly use x, no need for ast.literal_eval
                  (np.linalg.norm(query_embedding) * np.linalg.norm(x))  # Directly use x, no need for ast.literal_eval
    )
    top_k_indices = knowledge_df.sort_values(by="Similarity", ascending=False).head(2)["Index"].tolist()

    # Check if ground truth is in the top-2 and top-1
    is_in_top_2 = int(ground_truth_index in top_k_indices)
    is_in_top_1 = int(ground_truth_index == top_k_indices[0])

    # Save results
    retrieval_results.append({
        "Query": row["User_Question"],
        "Document_Retrived_1": knowledge_df.loc[top_k_indices[0], "Document"],
        "Document_Retrived_2": knowledge_df.loc[top_k_indices[1], "Document"],
        "Index_Document_Ground_Truth": ground_truth_index,
        "Indexes_Document_Retrived": top_k_indices,
        "Is_Ground_Truth_In_Top_2": is_in_top_2,
        "Is_Ground_Truth_In_Top_1": is_in_top_1
    })

# Convert results to DataFrame and save as CSV
results_df = pd.DataFrame(retrieval_results)
results_df.to_csv("retrival_result_document.csv", index=False)
print("Retrieval results saved to 'retrival_result_document.csv'.")

Retrieval results saved to 'retrival_result_document.csv'.


**Calculate Precision, Recall, and F1 Score**

In [ ]:
from sklearn.metrics import precision_score, recall_score, f1_score

# Extract true labels and predictions for both Top-2 and Top-1
y_true_top_2 = results_df["Is_Ground_Truth_In_Top_2"]
y_pred_top_2 = [1] * len(results_df)  # Since we always retrieve 2 documents, predictions are always 1

y_true_top_1 = results_df["Is_Ground_Truth_In_Top_1"]
y_pred_top_1 = [1] * len(results_df)  # Predictions are 1 as we evaluate Top-1 retrieval.

# Metrics for Top-2
precision_top_2 = precision_score(y_true_top_2, y_pred_top_2, zero_division=1)
recall_top_2 = recall_score(y_true_top_2, y_pred_top_2, zero_division=1)
f1_top_2 = f1_score(y_true_top_2, y_pred_top_2, zero_division=1)

# Metrics for Top-1
precision_top_1 = precision_score(y_true_top_1, y_pred_top_1, zero_division=1)
recall_top_1 = recall_score(y_true_top_1, y_pred_top_1, zero_division=1)
f1_top_1 = f1_score(y_true_top_1, y_pred_top_1, zero_division=1)

# Print results
print(f"Top-2 Precision: {precision_top_2}, Recall: {recall_top_2}, F1 Score: {f1_top_2}")
print(f"Top-1 Precision: {precision_top_1}, Recall: {recall_top_1}, F1 Score: {f1_top_1}")

Top-2 Precision: 0.9048582995951417, Recall: 1.0, F1 Score: 0.9500531349628055
Top-1 Precision: 0.8825910931174089, Recall: 1.0, F1 Score: 0.9376344086021505


**Analysis:** Precision has increased! This is dual-endcoder approach is better than no technique.

# Stage 2.2: Change RAG Retrival Strategy - retrive with different types of qustions

## Data preprocessing

In [ ]:
import pandas as pd
import ast

# Function to process columns (remove unwanted characters and clean up)
def process_column(value):
    if pd.isna(value):
        return None
    try:
        items = ast.literal_eval(value)
        return [item.strip() for item in items if item.strip()]
    except:
        return None

# Function to generate questions and answers for each ingredient
def generate_questions_answers(name, good_for, bad_for):
    qa_pairs = []
    if good_for:
        good_for_sentence = f"{name} is suitable for: {', '.join(good_for)}."
        qa_pairs.append({
            "Question": f"Who is {name} good for?",
            "Answer": good_for_sentence
        })
    if bad_for:
        bad_for_sentence = f"People with {', '.join(bad_for)} should avoid {name}."
        qa_pairs.append({
            "Question": f"Who should avoid {name}?",
            "Answer": bad_for_sentence
        })
    qa_pairs.append({
        "Question": f"What are the benefits of {name}?",
        "Answer": f"The benefits of {name} are outlined in the product description."
    })
    qa_pairs.append({
        "Question": f"How should {name} be used?",
        "Answer": f"{name} should be used as recommended by a professional or product instructions."
    })
    return qa_pairs

# Load your dataset
df = pd.read_csv('ingredientList_processed.csv')

# Clean the "who_is_it_good_for" and "who_should_avoid" columns
df['who_is_it_good_for'] = df['who_is_it_good_for'].apply(process_column)
df['who_should_avoid'] = df['who_should_avoid'].apply(process_column)

# Generate questions and answers for each ingredient
qa_pairs = []
for _, row in df.iterrows():
    name = row['name']
    good_for = row['who_is_it_good_for']
    bad_for = row['who_should_avoid']
    qa_pairs.extend(generate_questions_answers(name, good_for, bad_for))

# Convert to a DataFrame
qa_df = pd.DataFrame(qa_pairs)

# Save the questions and answers to a new CSV
qa_df.to_csv('ingredient_qa.csv', index=False)

print("Questions and answers successfully generated and saved to 'ingredient_qa.csv'.")

Questions and answers successfully generated and saved to 'ingredient_qa.csv'.


## Generate Embedding

In [ ]:
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd

In [ ]:
# Load the generated QA pairs
qa_df = pd.read_csv('ingredient_qa.csv')

#Step 1: Read the documents (preprocessed text)
documents=qa_df['Answer']
# Read ingredient names
question_encoder = qa_df['Question']

# Step 2: Create embeddings for the questions
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')  # Use a lightweight embedding model
question_embeddings = embedding_model.encode(question_encoder)

# Step 3: Create a pandas DataFrame for indexing
# Combine documents and embeddings in a DataFrame
df = pd.DataFrame({
    "Question_Encoder": question_encoder,
    "Document": documents,
    "Embedding": list(question_embeddings)  # Store embeddings as lists for easier handling
})

# Display the DataFrame
print("Indexed Documents:")
print(df.head())

Indexed Documents:
                                    Question_Encoder  \
0      Who is Alpha-Glucan Oligosaccharide good for?   
1     Who should avoid Alpha-Glucan Oligosaccharide?   
2  What are the benefits of Alpha-Glucan Oligosac...   
3   How should Alpha-Glucan Oligosaccharide be used?   
4                         Who is Aloe Vera good for?   

                                            Document  \
0  Alpha-Glucan Oligosaccharide is suitable for: ...   
1  People with Related Allergy should avoid Alpha...   
2  The benefits of Alpha-Glucan Oligosaccharide a...   
3  Alpha-Glucan Oligosaccharide should be used as...   
4  Aloe Vera is suitable for: Dry and dehydrated ...   

                                           Embedding  
0  [-0.01807728, -0.0152930245, -0.06922731, 0.05...  
1  [0.03905549, -0.02278369, -0.049849454, 0.0193...  
2  [-0.011615788, 0.015339728, -0.029196002, 0.04...  
3  [0.0063390145, -0.042018417, -0.08773068, 0.00...  
4  [-0.005596607, 0.07699273, 0.

In [ ]:
qa_df

Question  \
0        Who is Alpha-Glucan Oligosaccharide good for?   
1       Who should avoid Alpha-Glucan Oligosaccharide?   
2    What are the benefits of Alpha-Glucan Oligosac...   
3     How should Alpha-Glucan Oligosaccharide be used?   
4                           Who is Aloe Vera good for?   
..                                                 ...   
984                        How should Taurine be used?   
985         Who is Sodium Ascorbyl Phosphate good for?   
986        Who should avoid Sodium Ascorbyl Phosphate?   
987  What are the benefits of Sodium Ascorbyl Phosp...   
988      How should Sodium Ascorbyl Phosphate be used?   

                                                Answer  
0    Alpha-Glucan Oligosaccharide is suitable for: ...  
1    People with Related Allergy should avoid Alpha...  
2    The benefits of Alpha-Glucan Oligosaccharide a...  
3    Alpha-Glucan Oligosaccharide should be used as...  
4    Aloe Vera is suitable for: Dry and dehydrated ...  
..                                                 ...  
984  Taurine should be used as recommended by a pro...  
985  Sodium Ascorbyl Phosphate is suitable for: Acn...  
986  People with Related Allergy, Impaired skin bar...  
987  The benefits of Sodium Ascorbyl Phosphate are ...  
988  Sodium Ascorbyl Phosphate should be used as re...  

[989 rows x 2 columns]

In [ ]:
#Create a new file to save the embedding
df.to_csv('knowledge_embedding_qa.csv')

## Retrival Evaluation 2.2

In [ ]:
import ast  # Import ast to handle string representations of lists

test_df = pd.read_csv("Retrival_Test_Set_Embedding.csv")

# Convert string embeddings to lists
test_df['Query_Embedding'] = test_df['Query_Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split(",")],
        dtype=np.float32
    )
)

In [ ]:
# Load the knowledage data
knowledge_df=pd.read_csv('knowledge_embedding_qa.csv')

# Convert string embeddings to lists
import ast
knowledge_df['Embedding'] = knowledge_df['Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split()],
        dtype=np.float32
    )
)

print(type( knowledge_df['Embedding'][0]))

<class 'numpy.ndarray'>


In [ ]:
# Retrieve based on a question
query  = "Is Ascorbic Acid good for dry skin?"
results = retrieve_with_pandas(query, top_k=2)

print("\nQuery:", query)
print("\nTop Retrieved Documents:")
print(results)


Top Retrieved Documents:

Query: Is Ascorbic Acid good for dry skin?

Top Retrieved Documents:
                                              Document  Similarity
60   Ascorbic Acid is suitable for: Elasticity, Fin...    0.755207
458  L-Ascorbic Acid is suitable for: Elasticity, F...    0.703898


**Qualitative Analysis:** It has also retrived the right document. The difference is that this document is shorter and more specific.


Now let's do a quantitative test.

In [ ]:
import ast  # Import ast to handle string representations of lists

test_df = pd.read_csv("Retrival_Test_Set_Embedding.csv")

# Convert string embeddings to lists
test_df['Query_Embedding'] = test_df['Query_Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split(",")],
        dtype=np.float32
    )
)

In [ ]:
retrieval_results = []

for _, row in test_df.iterrows():
    query_embedding = np.array(row["Query_Embedding"])  # Load query embedding
    ground_truth_index = row["Index_Retrival_Question"] # The ground truth index

    # Compute similarity and retrieve top-2 documents
    knowledge_df['Similarity'] = knowledge_df['Embedding'].apply(
        lambda x: np.dot(query_embedding, x) / # Directly use x, no need for ast.literal_eval
                  (np.linalg.norm(query_embedding) * np.linalg.norm(x))  # Directly use x, no need for ast.literal_eval
    )
    top_k_indices = knowledge_df.sort_values(by="Similarity", ascending=False).head(2)["Index"].tolist()

    # Check if ground truth is in the top-2 and top-1
    is_in_top_2 = int(ground_truth_index in top_k_indices)
    is_in_top_1 = int(ground_truth_index == top_k_indices[0])

    # Save results
    retrieval_results.append({
        "Query": row["User_Question"],
        "Document_Retrived_1": knowledge_df.loc[top_k_indices[0], "Document"],
        "Document_Retrived_2": knowledge_df.loc[top_k_indices[1], "Document"],
        "Index_Document_Ground_Truth": ground_truth_index,
        "Indexes_Document_Retrived": top_k_indices,
        "Is_Ground_Truth_In_Top_2": is_in_top_2,
        "Is_Ground_Truth_In_Top_1": is_in_top_1
    })

# Convert results to DataFrame and save as CSV
results_df = pd.DataFrame(retrieval_results)
results_df.to_csv("retrival_result_document.csv", index=False)
print("Retrieval results saved to 'retrival_result_document.csv'.")

Retrieval results saved to 'retrival_result_document.csv'.


Calculate Precision, Recall, and F1 Score

In [ ]:
from sklearn.metrics import precision_score, recall_score, f1_score

# Extract true labels and predictions for both Top-2 and Top-1
y_true_top_2 = results_df["Is_Ground_Truth_In_Top_2"]
y_pred_top_2 = [1] * len(results_df)  # Since we always retrieve 2 documents, predictions are always 1

y_true_top_1 = results_df["Is_Ground_Truth_In_Top_1"]
y_pred_top_1 = [1] * len(results_df)  # Predictions are 1 as we evaluate Top-1 retrieval.

# Metrics for Top-2
precision_top_2 = precision_score(y_true_top_2, y_pred_top_2, zero_division=1)
recall_top_2 = recall_score(y_true_top_2, y_pred_top_2, zero_division=1)
f1_top_2 = f1_score(y_true_top_2, y_pred_top_2, zero_division=1)

# Metrics for Top-1
precision_top_1 = precision_score(y_true_top_1, y_pred_top_1, zero_division=1)
recall_top_1 = recall_score(y_true_top_1, y_pred_top_1, zero_division=1)
f1_top_1 = f1_score(y_true_top_1, y_pred_top_1, zero_division=1)

# Print results
print(f"Top-2 Precision: {precision_top_2}, Recall: {recall_top_2}, F1 Score: {f1_top_2}")
print(f"Top-1 Precision: {precision_top_1}, Recall: {recall_top_1}, F1 Score: {f1_top_1}")

Top-2 Precision: 0.8178137651821862, Recall: 1.0, F1 Score: 0.8997772828507795
Top-1 Precision: 0.5728744939271255, Recall: 1.0, F1 Score: 0.7284427284427284


**Analysis:** Precision has lowered! Top 1 precision is even worse than Attempt 1. This is dual-endcoder approach is not ideal, probably because there's more documents in the knowledge based now, so there's higher chance to mess up.

# Stage 2 Coclusion - Retrival Improvement


*   Attempt 2.1 sees signigicant improvement in precision. This might be because each ingredient is identical, so it's hard to mess up.

*   Attempt 2.2 has relatively lower precision, especially for Top-1. This might be because there's more documents in the knowledge based now, so there's higher chance to mess up. However, it's too early to say Attempt 2.2 is not helpful at all, because its document is shorter and tailored to specific questions, which might increase effectiveness in the response generation step.



# Stage 3.1 Generation Evaluation

### Test 3.0 Generate with no context:

Test 3.0.1:

In [16]:
query  = "Is Ascorbic Acid good for dry skin?"
response = llama_pipeline(
          query,
          do_sample=True,
          top_k=10,
          num_return_sequences=1,
          eos_token_id=tokenizer.eos_token_id,
          max_length=256,
          truncation=True
        )

In [19]:
print (response[0]['generated_text'])

Is Ascorbic Acid good for dry skin?

Ascorbic Acid, also known as Vitamin C, is a popular ingredient in skincare products that can provide numerous benefits for dry skin. Here are some reasons why Ascorbic Acid is good for dry skin:

1. Hydration: Ascorbic Acid can help to hydrate dry skin by increasing the production of glycosaminoglycans (GAGs), which are important components of the skin's natural moisture barrier.
2. Anti-inflammatory properties: Ascorbic Acid has anti-inflammatory properties that can help to reduce redness and irritation in dry skin.
3. Improves skin elasticity: Ascorbic Acid can help to improve skin elasticity, which can make dry skin appear more youthful and radiant.
4. Protects against environmental stressors: Ascorbic Acid can help to protect the skin against environmental stressors such as pollution, UV rays, and cold weather, which can exacerbate dry skin.
5. Fades


### Test 3.1 Generate with document retrived by ingredient name.

In [9]:
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Load the knowledage data

In [104]:
# Load the knowledage data
# !!!!! important !!!!!
knowledge_df=pd.read_csv('knowledge_embedding_name.csv')

# Convert string embeddings to lists
import ast
knowledge_df['Embedding'] = knowledge_df['Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split()],
        dtype=np.float32
    )
)

print(type( knowledge_df['Embedding'][0]))

<class 'numpy.ndarray'>


**Editted the function to retrieve knowledge, since the retrival percision using the ingredient name is quite high, we only take the top-1 document as context.**

In [105]:
#function to retrieve knowledge, only take the top 1 as context
def retrieve_with_pandas(query, top_k=1):
    # Generate embedding for the query
    query_embedding = embedding_model.encode([query])[0]

    # Compute similarity scores (cosine similarity)
    knowledge_df['Similarity'] =  knowledge_df['Embedding'].apply(lambda x: np.dot(query_embedding, x) /
                                             (np.linalg.norm(query_embedding) * np.linalg.norm(x)))
    # Sort by similarity and return top-k results
    results =  knowledge_df.sort_values(by="Similarity", ascending=False).head(top_k)
    #print("\nTop Retrieved Documents:") # for debugging

    return results[["Document", "Similarity"]]

Set up the pipeline.

In [8]:
# Set up pipeline
llama_pipeline = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer,
    device_map="auto"
)

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

The function for RAG

In [109]:
#Debug supported by Chat-GPT

def generate_with_rag_Llama(query):
    # Retrieve context
    results = retrieve_with_pandas(query)

    # Check if there's any retrieved context
    if not results.empty:
        context = results["Document"].iloc[0].strip("\n").strip("\r")   # Get the top retrieved document
        # Combine context with query
        input_text = f"Respond the query based on the context.Context: {context}. Query: {query} \n"
        print("Generating with Input Prompt Text: "+input_text)
        print("Result: ")
        # Use LLM to generate a response
        response = llama_pipeline(
          input_text,
          do_sample=True,
          top_k=10,
          num_return_sequences=1,
          eos_token_id=tokenizer.eos_token_id,
          max_length=512,
          truncation=True
        )
        return response[0]['generated_text']
    else:
        print("No relevant context found.")

Test 3.1.1

In [110]:
query  = "Is Ascorbic Acid good for dry skin?"
response= generate_with_rag_Llama(query)
response

Generating with Input Prompt Text: Respond the query based on the context.Context:  Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.
 People with Related Allergy,Sensitive,Dry Dehydrated,Impaired skin barrier should avoid Ascorbic Acid.
Ascorbic Acid: Ascorbic acid is a powerhouse ingredient with evidence for collagen synthesis so can help make the skin firmer, helps with hyperpigmentation, dark spots, and uneven skin tone and is great for preventing damage from the sun by neutralising free radicals.

Since ascorbic acid is already in the active form needed by the skin, no conversion is necessary, and if stabilised is highly potent.. Query: Is Ascorbic Acid good for dry skin? 

Result: 


"Respond the query based on the context.Context:  Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.\n People with Related Allergy,Sensitive,Dry Dehydrated,Impaired skin barrier should avoid Ascorbic Acid.\nAscorbic Acid: Ascorbic acid is a powerhouse ingredient with evidence for collagen synthesis so can help make the skin firmer, helps with hyperpigmentation, dark spots, and uneven skin tone and is great for preventing damage from the sun by neutralising free radicals.\r\n\r\nSince ascorbic acid is already in the active form needed by the skin, no conversion is necessary, and if stabilised is highly potent.. Query: Is Ascorbic Acid good for dry skin? \n\nAnswer: Ascorbic acid is suitable for elasticity, fine lines, pigmentation, radiance, pregnancy, and wrinkles. However, people with related allergies, sensitivities, dry dehydrated, or impaired skin barrier should avoid ascorbic acid.\n\nAscorbic acid is not recommended for dry skin. A

**Full Prompt:**
> Respond the query based on the context.Context:  Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.
> People with Related Allergy,Sensitive,Dry Dehydrated,Impaired skin barrier should avoid Ascorbic Acid.
Ascorbic Acid: Ascorbic acid is a powerhouse ingredient with evidence for collagen synthesis so can help make the skin firmer, helps with hyperpigmentation, dark spots, and uneven skin tone and is great for preventing damage from the sun by neutralising free radicals.
> Since ascorbic acid is already in the active form needed by the skin, no conversion is necessary, and if stabilised is highly potent.. Query: Is Ascorbic Acid good for dry skin?


---

**Full Answer:**
> Ascorbic acid is suitable for elasticity, fine lines, pigmentation, radiance, pregnancy, and wrinkles. However, people with related allergies, sensitivities, dry dehydrated, or impaired skin barrier should avoid ascorbic acid.

> Ascorbic acid is not recommended for dry skin. Ascorbic acid is a powerful antioxidant that can help neutralize free radicals and improve skin health, but it can also be drying and irritating to dry skin. It is best to use ascorbic acid in small amounts and to moisturize the skin before applying it.

> It is important to note that ascorbic acid can be irritating to dry or sensitive skin, so it's best to start with a small amount and gradually increase as needed. It's also important to use a moisturizer before applying ascorbic acid to help protect the skin and prevent irritation.

**Analysis:** The answer is more relevant and is in the right direction.

### Test 3.2 Generate with document retrived by example questions.

In [ ]:
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Load the knowledage data

In [112]:
# Load the knowledage data
# !!!!! important !!!!!
knowledge_df=pd.read_csv('knowledge_embedding_qa.csv')

# Convert string embeddings to lists
import ast
knowledge_df['Embedding'] = knowledge_df['Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split()],
        dtype=np.float32
    )
)

print(type( knowledge_df['Embedding'][0]))

<class 'numpy.ndarray'>


**Editted the function to retrieve knowledge, since the retrival percision using the sample question is lower, we will take the top-2 documents as context.**

In [121]:
#function to retrieve knowledge, take the top 2 as context
def retrieve_with_pandas(query, top_k=2):
    # Generate embedding for the query
    query_embedding = embedding_model.encode([query])[0]

    # Compute similarity scores (cosine similarity)
    knowledge_df['Similarity'] =  knowledge_df['Embedding'].apply(lambda x: np.dot(query_embedding, x) /
                                             (np.linalg.norm(query_embedding) * np.linalg.norm(x)))
    # Sort by similarity and return top-k results
    results =  knowledge_df.sort_values(by="Similarity", ascending=False).head(top_k)
    #print("\nTop Retrieved Documents:") # for debugging

    return results[["Document", "Similarity"]]

In [122]:
retrieve_with_pandas(query)

Document  Similarity
60   Ascorbic Acid is suitable for: Elasticity, Fin...    0.755207
458  L-Ascorbic Acid is suitable for: Elasticity, F...    0.703898

Set up the pipeline.

In [ ]:
# Set up pipeline
llama_pipeline = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer,
    device_map="auto"
)

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

The function for RAG

In [124]:
#Debug supported by Chat-GPT

def generate_with_rag_Llama(query):
    # Retrieve context
    results = retrieve_with_pandas(query)

    # Check if there's any retrieved context
    if not results.empty:
        # Concatenate all "Document" values into a single string
        context = "\n".join(results["Document"].str.strip("\n").str.strip("\r"))
        # Combine context with query
        input_text = f"Respond the query based on the context.Context: {context}. Query: {query} \n"
        print("Generating with Input Prompt Text: "+input_text)
        print("Result: ")
        # Use LLM to generate a response
        response = llama_pipeline(
          input_text,
          do_sample=True,
          top_k=10,
          num_return_sequences=1,
          eos_token_id=tokenizer.eos_token_id,
          max_length=512,
          truncation=True
        )
        return response[0]['generated_text']
    else:
        print("No relevant context found.")

*Test* 3.2.1

In [113]:
query  = "Is Ascorbic Acid good for dry skin?"
response= generate_with_rag_Llama(query)
response

Generating with Input Prompt Text: Respond the query based on the context.Context: Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.. Query: Is Ascorbic Acid good for dry skin? 

Result: 


"Respond the query based on the context.Context: Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.. Query: Is Ascorbic Acid good for dry skin? \nAnswer: Ascorbic Acid is not suitable for dry skin. While it is effective in addressing various skin concerns such as elasticity, fine lines, pigmentation, radiance, and pregnancy, it may not be the best option for dry skin. Ascorbic acid can be drying and irritating to the skin, especially if used in high concentrations or for extended periods. It's important to choose a gentler, hydrating skincare product that won't exacerbate dryness. Instead, look for products that contain humectants such as hyaluronic acid, glycerin, or ceramides, which help lock in moisture and soothe dryness."

**Full Prompt:**
> Respond the query based on the context.Context: Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.. Query: Is Ascorbic Acid good for dry skin?


---


**Full Answer:**

> Ascorbic Acid is not suitable for dry skin. While it is effective in addressing various skin concerns such as elasticity, fine lines, pigmentation, radiance, and pregnancy, it may not be the best option for dry skin. Ascorbic acid can be drying and irritating to the skin, especially if used in high concentrations or for extended periods. It's important to choose a gentler, hydrating skincare product that won't exacerbate dryness. Instead, look for products that contain humectants such as hyaluronic acid, glycerin, or ceramides, which help lock in moisture and soothe dryness.

**Analysis:**

More concise. More relevant. Generation is faster comapare to Test 3.1, because context is shorter, less to process. This is better for real-time interaction.

* Generation time for Test 3.1.1: 8 seconds.

* Generation time for Test 3.2.1: 5 seconds.



# Stage 3.2 Generation Improvement

### Prompt Engnieering 1

The answer is still a bit long. Let's change the prompt from
- "Respond the query based on the context.Context: {context}. Query: {query} \n"

to
- "Answer the query concisely using the context.\nContext: {context}\nQuery: {query}\n"

see if it will do better.

In [128]:
def generate_with_rag_Llama(query):
    # Retrieve context
    results = retrieve_with_pandas(query)

    # Check if there's any retrieved context
    if not results.empty:
        # Concatenate all "Document" values into a single string
        context = "\n".join(results["Document"].str.strip("\n").str.strip("\r"))
        # Combine context with query
        input_text = f"Answer the query concisely using the context.\nContext: {context}\nQuery: {query}\n"
        print("Generating with Input Prompt Text: "+input_text)
        print("Result: ")
        # Use LLM to generate a response
        response = llama_pipeline(
          input_text,
          do_sample=True,
          top_k=10,
          num_return_sequences=1,
          eos_token_id=tokenizer.eos_token_id,
          max_length=512,
          truncation=True
        )
        return response[0]['generated_text']
    else:
        print("No relevant context found.")

In [129]:
query  = "Is Ascorbic Acid good for dry skin?"
response= generate_with_rag_Llama(query)
response

Generating with Input Prompt Text: Answer the query concisely using the context.
Context: Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.
L-Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.
Query: Is Ascorbic Acid good for dry skin?

Result: 


'Answer the query concisely using the context.\nContext: Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.\nL-Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.\nQuery: Is Ascorbic Acid good for dry skin?\nAnswer: Ascorbic Acid is not suitable for dry skin. L-Ascorbic Acid is suitable for dry skin.'


**Answer:**
> Ascorbic Acid is not suitable for dry skin. L-Ascorbic Acid is suitable for dry skin.

**Analysis:**

This might be too short. We still want some level of the context in the output to justify the answer.



### Prompt Engnieering 2

Let's change the prompt from
- "Answer the query concisely using the context.\nContext: {context}\nQuery: {query}\n"

to
- "Answer the query concisely using the context to provide reasoning.\nContext: {context}\nQuery: {query}\n"

see if it will do better.

In [137]:
def generate_with_rag_Llama(query):
    # Retrieve context
    results = retrieve_with_pandas(query)

    # Check if there's any retrieved context
    if not results.empty:
        # Concatenate all "Document" values into a single string
        context = "\n".join(results["Document"].str.strip("\n").str.strip("\r"))
        # Combine context with query
        input_text = f"Answer the query concisely using the context to provide reasoning.\nContext: {context}\nQuery: {query}\n"
        print("Generating with Input Prompt Text: "+input_text)
        print("Result: ")
        # Use LLM to generate a response
        response = llama_pipeline(
          input_text,
          do_sample=True,
          top_k=10,
          num_return_sequences=1,
          eos_token_id=tokenizer.eos_token_id,
          max_length=512,
          truncation=True
        )
        return response[0]['generated_text']
    else:
        print("No relevant context found.")

In [139]:
query  = "Is Ascorbic Acid good for dry skin?"
response= generate_with_rag_Llama(query)
response

Generating with Input Prompt Text: Answer the query concisely using the context to provide reasoning.
Context: Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.
L-Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.
Query: Is Ascorbic Acid good for dry skin?

Result: 


'Answer the query concisely using the context to provide reasoning.\nContext: Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.\nL-Ascorbic Acid is suitable for: Elasticity, Fine Lines, Pigmentation, Radiance, Pregnancy, Wrinkles.\nQuery: Is Ascorbic Acid good for dry skin?\nAscorbic Acid is not suitable for dry skin. Ascorbic acid is a humectant, meaning it attracts and retains moisture. However, it may not be the best option for dry skin as it may exacerbate the dryness. L-Ascorbic Acid, on the other hand, is suitable for dry skin as it is a more gentle form of ascorbic acid that is less likely to irritate the skin.'

**Answer:**


> Acid is not suitable for dry skin. Ascorbic acid is a humectant, meaning it attracts and retains moisture. However, it may not be the best option for dry skin as it may exacerbate the dryness. L-Ascorbic Acid, on the other hand, is suitable for dry skin as it is a more gentle form of ascorbic acid that is less likely to irritate the skin.

**Analysis**

This looks better and more intellegent because it can reason like human.

# Final - Try It Out :)

Run all the code below to try out the chatbot

Log in to Hugging Face and Load Llama-2 and tokenizer

In [ ]:
from huggingface_hub import login

login(token="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")
from transformers import AutoTokenizer
import transformers
import torch

model = "meta-llama/Llama-2-7b-chat-hf" # meta-llama/Llama-2-7b-hf

tokenizer = AutoTokenizer.from_pretrained(model, use_auth_token=True)

Prepare the embedding model

In [150]:
from transformers import pipeline
from sentence_transformers import SentenceTransformer
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')

import pandas as pd
import numpy as np

Load the knowledage data

In [ ]:
# Load the knowledage data
# !!!!! important !!!!!
knowledge_df=pd.read_csv('knowledge_embedding_name.csv')

# Convert string embeddings to lists
import ast
knowledge_df['Embedding'] = knowledge_df['Embedding'].apply(
    lambda x: np.array(
        [float(num) for num in x.strip("[]").split()],
        dtype=np.float32
    )
)

print(type( knowledge_df['Embedding'][0]))

<class 'numpy.ndarray'>


In [153]:
#function to retrieve knowledge, take the top 1 as context
def retrieve_with_pandas(query, top_k=2):
    # Generate embedding for the query
    query_embedding = embedding_model.encode([query])[0]

    # Compute similarity scores (cosine similarity)
    knowledge_df['Similarity'] =  knowledge_df['Embedding'].apply(lambda x: np.dot(query_embedding, x) /
                                             (np.linalg.norm(query_embedding) * np.linalg.norm(x)))
    # Sort by similarity and return top-k results
    results =  knowledge_df.sort_values(by="Similarity", ascending=False).head(top_k)
    #print("\nTop Retrieved Documents:") # for debugging

    return results[["Document", "Similarity"]]

Set up the pipeline.

In [ ]:
# Set up pipeline
llama_pipeline = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer,
    device_map="auto"
)

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

The function for RAG

In [160]:
#Debug supported by Chat-GPT

def generate_with_rag_Llama(query):
    # Retrieve context
    results = retrieve_with_pandas(query)

    # Check if there's any retrieved context
    if not results.empty:
        # Concatenate all "Document" values into a single string
        context = "\n".join(results["Document"].str.strip("\n").str.strip("\r"))
        # Combine context with query
        input_text = f"Answer the query concisely using the context to provide reasoning.\nContext: {context}\nQuery: {query}\n"
        #print("Generating with Input Prompt Text: "+input_text)
        #print("Result: ")
        # Use LLM to generate a response
        response = llama_pipeline(
          input_text,
          do_sample=True,
          top_k=10,
          num_return_sequences=1,
          eos_token_id=tokenizer.eos_token_id,
          max_length=512,
          truncation=True
        )
        return response[0]['generated_text'].split(input_text)[-1] # This removes the prompt from the response
    else:
        print("No relevant context found.")

Put things into a loop.

**1. Run the code to start typing question.**

**2. Hit "Enter" on keyboard to get response.**

**3. Type "bye", "quit", "exit", to exit!**




In [159]:
while True:
    user_input = input("You: ")
    if user_input.lower() in ["bye", "quit", "exit"]:
        print("Chatbot: Goodbye!")
        break
    print("Chatbot:", generate_with_rag_Llama(user_input))

You: Should people with dry skin use Ascorbic Acid?
Chatbot: 
Answer: Ascorbic Acid is not suitable for people with dry skin. Ascorbic Acid can exacerbate dryness and irritation, leading to further skin barrier impairment. Individuals with dry skin, sensitive skin, or impaired skin barrier should avoid using Ascorbic Acid. Instead, they can opt for alternative vitamin C sources that are gentler on the skin, such as liposomal vitamin C or vitamin C esters. These forms of vitamin C are more easily absorbed and less likely to cause irritation, making them a better option for individuals with dry or sensitive skin.
You: bye
Chatbot: Goodbye!
